### DISCUSSION TITLE
How to reference a NuGet package (and its related dependencies) in PS7

### DISCUSSION THREAD

**Discussion by netcorefan1** [PowerShell/PowerShell#19722] [Q&A]
Hello,
I have a C# library (which reference some NuGet packages) that I want to share with PS scripts. For easiness I just build it as .nuPkg package which I move into my local repo. When trying to use `Install-Package` I get any sort of troubles, from missing dependencies to type not found.
It seems to be that using NuGet packages in PS Core is very [problematic](https://stackoverflow.com/questions/39257572/loading-assemblies-from-nuget-packages) and I would like to know what would be, as of now, the best way to handle this.

**Comment by doctordns** [comment]
What precisely are you doing?

**Comment by rhubarb-geek-nz** [comment]
The [Install-Package](https://learn.microsoft.com/en-us/powershell/module/packagemanagement/install-package) lists the following package types

- Cmdlet
- DscResource
- Function
- RoleCapability
- Workflow

Which suggests it only supports installing that type of package. I don't know what the difference between Install-Package and [Install-Module](https://learn.microsoft.com/en-us/powershell/module/powershellget/install-module) is, I have only used the latter.

So the question is what are you expecting to happen, where will the files be put and how will they be used? For example installing a module puts the files into a directory on the $Env:PSModulePath where they can be found.

There is a big difference between a NuGet package generated by "dotnet pack" and that produced by [Publish-Module](https://learn.microsoft.com/en-us/powershell/module/powershellget/publish-module).

If your library was packaged up as a Cmdlet for example, then it could be used once it is imported into the process. You would have a problem of versioning if you had the same library within multiple different Cmdlet modules.

**Reply by netcorefan1** [reply]
Thanks for your help. I really appreciate.

After I moved the VS2022 compiled NuGet package in `C:\MyLocalRepo`, I can install it with:
`Install-Package -Name MyPackageName -Scope CurrentUser -Source C:\MyLocalRepo -Force`
I need to first install the package dependencies, otherwise I get `Install-Package: Unable to find dependent package(s) (Serilog)`.
Shouldn't be automatically installed from the NuGet repo?

Packages are installed in `C:\Users\User\AppData\Local\PackageManagement\NuGet` which is something I would like to avoid because it creates unneeded duplicates. Any chance to get them installed in the default NuGet repo used by VS (`C:\Users\user\.nuget`) ?

Once done, PWSH is unable to find the the assemblies and I'm forced to manually specify (with hard coded paths) the required assemblies in a manifest (psd1) or through `Add-Type -LiteralPath ...`.

It is a .Net 7 library compiled in VS with the option `Generate NuGet package on build`.  In VS there is the "Publish" command, but I couldn't find any option to package as a CmdLet. Is `Publish-Module` only for online NuGet-based gallery? I don't need any online gallery. The ability to install my package from my local repo and get it automatically referenced as I do with VS is enough.

**Comment by rhubarb-geek-nz** [comment]
The VS "%USERPROFILE%\\.nuget" is really only for developers, it is not expected to be used by end users, binaries are normally installed on the target system using either msi or appx. The nuget package generated by `dotnet pack` isn't really intended for end user deployment, it has multiple builds of the binaries depending on how many frameworks and platforms you have.

The PowerShell system is really designed for installing modules from a repository like PSGallery, which is a nuget repository.

I am not sure how one is supposed to build PowerShell modules, I build with a custom task added to the VS csproj that runs a PowerShell script which uses [New-ModuleManifest](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/new-modulemanifest). My script creates the directory that can then be published with Publish-Module.

I run a local NuGet server in a [docker](https://powershellexplained.com/2018-03-03-Powershell-Using-a-NuGet-server-for-a-PSRepository/)

However this is not the same as running the nuget.server.core, for instance it does not enforce licensing.

**Reply by netcorefan1** [reply]
Thanks again for your help.

> The VS "%USERPROFILE%\.nuget" is really only for developers, it is not expected to be used by end users, binaries are normally installed on the target system using either msi or appx. The nuget package generated by `dotnet pack` isn't really intended for end user deployment, it has multiple builds of the binaries depending on how many frameworks and platforms you have.
> 
> The PowerShell system is really designed for installing modules from a repository like PSGallery, which is a nuget repository.

I'm not sure if this can make any difference, but I'm the only one developer and the only one end user at the same time. To better think on this, I really shouldn't care if I end up with the same duplicated binaries in two different locations.

> 
> I am not sure how one is supposed to build PowerShell modules, I build with a custom task added to the VS csproj that runs a PowerShell script which uses [New-ModuleManifest](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/new-modulemanifest). My script creates the directory that can then be published with Publish-Module.

I also used `New-ModuleManifest` and this is acceptable, but how you have been able to add paths without hard coding?
For example:
```
RequiredAssemblies = @("C:\Users\User\AppData\Local\PackageManagement\NuGet\mylibrary1.0.0\lib\net7.0\mylibrary.dll",
                       "C:\Users\User\AppData\Local\PackageManagement\NuGet\Serilog.3.0.0-dev-02012\lib\net7.0\Serilog.dll")
```

1. Is there an environment variable I can put here in order to avoid `C:\Users\User\AppData\Local\PackageManagement\NuGet` ?
2. May be I can force VS to output the same path each time I compiles `mylibrary.dll`, but its others dependencies comes from an online NuGet repo. It will be very problematic to manually change the path each time a package gets updated and manually append a new path each time I add a new NuGet package to `mylibrary`. Is there a way to only add `mylibrary.dll` to `RequiredAssemblies` and get its related NuGet dependencies automatically referenced?
3. I used `nuget install mylibrary -prerelease` because this allows to also install the related dependencies. I have not been able to do the same with `Install-package` which forced me to install each package separately. Is this a limitation or I'm doing something wrong?

**Reply by ThomasNieto** [reply]
The standard practice for nuget dependencies is to ship a module with your cmdlets and have the dlls in your module and reference them using the relative path in the module manifest so they get auto loaded on module import.

**Reply by netcorefan1** [reply]
@ThomasNieto This sounds more feasible. If I have understood well, I should bundle all my PWSH modules in the NuGet package created from VS. Then, after an `Install-package` command I can access everything from within my script. Is that correct?
